>From 7094ed8c64ad5979314e16ef8e8c18e5dcd1c4bd Mon Sep 17 00:00:00 2001
From: Jochen Sprickerhof <jochen@sprickerhof.de>
Date: Wed, 25 Sep 2013 09:21:13 +0200
Subject: [PATCH] urlview

durlview taken from:
https://github.com/muennich/dotfiles/blob/master/bin/durlview
---
 config.def.h |  1 +
 durlview     | 16 ++++++++++++++++
 st.c         | 26 +++++++++++++++++++++++++-
 3 files changed, 42 insertions(+), 1 deletion(-)
 create mode 100755 durlview

diff --git a/config.def.h b/config.def.h
index def6c9e..c8621d1 100644
--- a/config.def.h
+++ b/config.def.h
@@ -112,6 +112,7 @@ static Shortcut shortcuts[] = {
 	{ ShiftMask,		XK_Insert,	selpaste,	{.i =  0} },
 	{ MODKEY|ShiftMask,	XK_Insert,	clippaste,	{.i =  0} },
 	{ MODKEY,		XK_Num_Lock,	numlock,	{.i =  0} },
+	{ MODKEY,		XK_u,		urlview,	{0} },
 };
 
 /*
diff --git a/durlview b/durlview
new file mode 100755
index 0000000..cdc31e1
--- /dev/null
+++ b/durlview
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+input=$({ while read line; do echo $line; done } | grep -Eo '(((http|https|ftp|gopher)|mailto)[.:][^	 >"]*|www\.[-a-z0-9.]+)[^	 .,;">\):]' | tac | dmenu -l 10 -i)
+
+case "$input" in
+  *\ *)
+    url=$(echo "$input" | cut -d " " -f 1)
+    echo -n "$url" | xclip
+    ;;
+  "")
+    break
+    ;;
+  *)
+    surf "$input" &
+    ;;
+esac
diff --git a/st.c b/st.c
index 50b58a7..a5ce5a5 100644
--- a/st.c
+++ b/st.c
@@ -311,6 +311,7 @@ typedef struct {
 static void clippaste(const Arg *);
 static void numlock(const Arg *);
 static void selpaste(const Arg *);
+static void urlview(void);
 static void xzoom(const Arg *);
 
 /* Config.h for applying patches and the configuration. */
@@ -1203,10 +1204,14 @@ execsh(void) {
 void
 sigchld(int a) {
 	int stat = 0;
+	pid_t p;
 
-	if(waitpid(pid, &stat, 0) < 0)
+	if((p = waitpid(pid, &stat, WNOHANG)) < 0)
 		die("Waiting for pid %hd failed: %s\n", pid, SERRNO);
 
+	if(pid != p)
+		return;
+
 	if(WIFEXITED(stat)) {
 		exit(WEXITSTATUS(stat));
 	} else {
@@ -1264,6 +1269,25 @@ dump(char c) {
 }
 
 void
+urlview(void) {
+	int x, y;
+	long s;
+	FILE *d;
+
+	d = popen("durlview", "w");
+
+	for(x = 0; x < term.row; x++) {
+		for(y = 0; y < term.col; y++) {
+			utf8decode(TLINE(x)[y].c, &s);
+			fprintf(d, "%lc", (wint_t)s);
+		}
+	}
+	fprintf(d, "\n");
+
+	pclose(d);
+}
+
+void
 ttyread(void) {
 	static char buf[BUFSIZ];
 	static int buflen = 0;
-- 
1.8.4.rc3

